### ISSUE TITLE
Allow for Partial Updating Objects

### ISSUE THREAD

**Issue by onebrightlight** [riasvdv/craft-scout#158]
Currently AFAICT Scout only allows full updating of objects in Algolia. I was wondering if we could instead update individual properties in an object as mentioned in [their docs](https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/).

Use case: Page view ranking is an important part of our indexes, but given the number of relationships and additional fields we have on each standard record, pushing a nightly or even weekly update of those page views would be costly in terms of operations. It would be better if we could just update the pageview attribute of every one of those records.

**Comment by lindseydiloreto** [comment]
This would be helpful for people who want to frequently update their Algolia records based on data from several of my plugins...
 - [View Count](https://github.com/doublesecretagency/craft-viewcount)
 - [Upvote](https://github.com/doublesecretagency/craft-upvote)
 - [Star Ratings](https://github.com/doublesecretagency/craft-starratings)

I could probably provide several other real-world use cases in addition to the one @onebrightlight mentioned. 🙂

**Comment by timkelty** [comment]
Yup. This would be nice.

FWIW, I do this kind of thing on a nightly cron using `partialUpdateObjects`. In this example I refresh a "random" property on every object, because the client wants a random sort that gets refreshed daily 🤷‍♀️.

```
<?php

namespace modules\appmodule\console\controllers;

use Algolia\AlgoliaSearch\SearchClient;
use Craft;
use rias\scout\Scout;
use Tightenco\Collect\Support\Collection;
use yii\console\ExitCode;
use yii\helpers\Console;

class AlgoliaController extends \craft\console\Controller
{
    public function actionRandomizeIndex(string $indexName): int
    {
        $scoutSettings = Scout::$plugin->getSettings();
        $scoutIndex = $scoutSettings->getIndices()
            ->firstWhere('indexName', $indexName);

        $index = Craft::$container->get(SearchClient::class)->initIndex($indexName);
        $ids = $scoutIndex->criteria->ids();
        $updates = Collection::make($ids)
            ->map(function ($id) use ($ids) {
                return [
                  'objectID' => $id,
                  'random' => rand(0, count($ids)),
                ];
            });

        $batches = Collection::make($index->partialUpdateObjects($updates->all()));
        $total = Collection::make($batches)->reduce(function ($carry, $item) {
            return $carry + count($item['objectIDs']);
        }, 0);

        $this->stdout("{$total} element(s) updated in {$batches->count()} batch(es).\n");

        return ExitCode::OK;
    }
}
```

Same kind of thing should work for ratings, etc. I would think, but it would be nice baked in.

@onebrightlight With your example, it seems like you'd still need to do this on a cron, right? Batching is what really saves you the operations, and partial updates just let you fit more in a batch.

**Comment by onebrightlight** [comment]
@timkelty yes, it’d still need to be a cronjob for me. Are you saying that it can’t be a partial or batch job in that case?

**Comment by timkelty** [comment]
Nope, just clarifying.

That's exactly what I'm doing in my example above (I run `craft appmodule/randomize-index products` on a cron).

**Comment by howells** [comment]
Just wondered if there was any movement on this one - I update popularity metrics of my algolia entries via another (node) script, which means that when an entry is saved, those attributes are simply overwritten and I'd need to wait for the script to run again before they're updated. I'd love Scout to just update the attributes defined in scout, leaving anything else in-tact. Would that be possible?

**Comment by andrewmenich** [comment]
I implemented this feature in a local fork of the plugin that we're using and would be happy to submit a PR. Is the `v5` branch for Craft 5 the only active branch or is Craft 4 still receiving updates as well? Thanks!

**Comment by janhenckens** [comment]
> I implemented this feature in a local fork of the plugin that we're using and would be happy to submit a PR. Is the `v5` branch for Craft 5 the only active branch or is Craft 4 still receiving updates as well? Thanks!

Going forward that would be `v5` and `develop-v5`- Happy to review a PR 🙏
